### ISSUE TITLE
Incorrect accented chars replacement for ö and ä

### ISSUE THREAD

**Issue by daresh** [PrestaShop/PrestaShop#26157]
#### Describe the bug

When importing a product with the above letters in name, the friendly URL is not perfectly correct, ö is changed to o, and ä is changed to a

#### Expected behavior

ö should be changed to oe and ä should be changed to ae

#### Steps to Reproduce

1. Prepare a simple CSV file with only one cell - a name of a product containing "ö ä" letters
2. Import the file
3. Check the friendly URL created
4. Compare that friendly URL to the one generated by the "regenerate" button in back office 

#### Additional information

* PrestaShop version: all 1.6 and 1.7
* PHP version: all versions

**Comment by khouloudbelguith** [comment]
Hi @daresh,

I checked product creation
1. `ö` converted to `oe`
![image](https://user-images.githubusercontent.com/16067358/136539898-b0ce3024-ff4e-4407-8535-5e6da8016b56.png)
2. `å` converted to `a`
![image](https://user-images.githubusercontent.com/16067358/136540015-29b44123-661e-4442-addc-6652e09c2365.png)
3. `ä` converted `ä`
![image](https://user-images.githubusercontent.com/16067358/136540100-96bbf19f-e640-4593-a8ad-05a49f481378.png)

I have the same behavior with PS1769 and PS1780 
With the import
![image](https://user-images.githubusercontent.com/16067358/136541120-d0f74ad4-2f7f-41b7-972f-8bd96f1c41ec.png)

- `Rewrite link for ö (ID null): re-written as o.
- Rewrite link for å (ID null): re-written as a.
- Rewrite link for ä (ID null): re-written as a.`

I have the same issue with PS1769 and PS1780 
![image](https://user-images.githubusercontent.com/16067358/136541756-209fbfc5-6e67-4fe1-a81b-922aacddf589.png)

I'll add this to the debug backlog so that it's fixed.

Please be aware that due to the high number of reported issues, some bugs might take a very long time to be resolved. If this issue is important to you and you cannot wait for it to be fixed on the project's own time, we strongly suggest you consider [hiring a specialist](https://www.prestashop.com/en/experts) to help you. 

If you manage to get the problem fixed, then please contribute the fix back to the project as a [Pull Request](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/). Remember that the more people contribute, the better PrestaShop becomes for everyone.

Thanks!

**Comment by khouloudbelguith** [comment]
@daresh, I updated the title and the first comment because `å converted to a` in the creation of the product.

Thanks!

**Comment by daresh** [comment]
Thanks, I'm planning to contribute on this issue, maybe I'll stick to ö and ä so it's consistent between the button and the import.

**Comment by matks** [comment]
Hello,

> ö should be changed to oe and ä should be changed to ae

Why 'oe' is better than 'o' 🤔  ? Is there a SEO rule to follow for such a usecase ? Is there a standard ?

**Comment by daresh** [comment]
Hello,

> Why 'oe' is better than 'o' 🤔 ? Is there a SEO rule to follow for such a usecase ? Is there a standard ?

Two factors here:
1. Backoffice JavaScript does it exactly this way, so it should be consistent between JS and PHP
2. I got that info from one of my austrian customers, who considered it a bug, so I assume it's some standard for the german speaking part of the world.

**Comment by matks** [comment]
> I got that info from one of my austrian customers, who considered it a bug, so I assume it's some standard for the german speaking part of the world.

Thanks for sharing, but there is something I am scared of 😱  : what if some people in the world expect 'o' and other 'oe' 😅  .

@Julievrz could you reach out to some translators from countries where ö is used (Austria, Germany...) and verify there is a consensus about this fact 😉  ? So we avoid doing this then having to revert it later 😅

**Comment by daresh** [comment]
Did you read point 1?

PrestaShop already does it exactly like this in the Back Office. Only the old PHP code has it done differently. Look at the screenshots above.

**Comment by matks** [comment]
I read point 1 but what if the JS is doing it the wrong way 😉  . Unfortunately we have a good quantity of old behaviors inherited from the past that are wrong.

So now that you raise the topic, and it's great, it's a good opportunity to explore what should be the right way to handle this usecase (now matter what is happening in the software nowadays), verify it works for kinda everyone and then implement the right solution once and for good 😉  . That's why I just want to check with multiple users if the above solution is generic and universal.

**Comment by daresh** [comment]
Allright, always good to verify this information with some native speakers :)

**Comment by matks** [comment]
@Julievrz hello 😄 

> could you reach out to some translators from countries where ö is used (Austria, Germany...) and verify there is a consensus about this fact 😉 ? So we avoid doing this then having to revert it later 😅

**Comment by davidglezz** [comment]
I think this solution is better, less code to maintain
https://github.com/PrestaShop/PrestaShop/pull/27176

**Comment by daresh** [comment]
@davidglezz with this solution there is still inconsistency. Different bahavior in JS, different in PHP. It should not be like that. If we talk about how PrestaShop handles accented chars, it should be the same for JS and for PHP.

**Comment by Julievrz** [comment]
> @Julievrz hello 😄
> 
> > could you reach out to some translators from countries where ö is used (Austria, Germany...) and verify there is a consensus about this fact 😉 ? So we avoid doing this then having to revert it later 😅

Sorry I completely missed the previous message. I'll contact some translators.

**Comment by davidglezz** [comment]
@daresh I think we should get rid of the JS function and make a request to the back-office, this way the result is always the same.

**Comment by icemansparks** [comment]
The example is correct, if (in German) you can’t use „umlauts“, we will fall back to the following:

ö -> oe / Ö -> Oe
ä -> ae / Ä -> Ae
ü -> ue / Ü -> Ue 
ẞ -> ss

**Comment by Julievrz** [comment]
> The example is correct, if (in German) you can’t use „umlauts“, we will fall back to the following:
> 
> ö -> oe / Ö -> Oe ä -> ae / Ä -> Ae ü -> ue / Ü -> Ue ẞ -> ss

Hi Dennis, thank you for answering!

**Comment by Julievrz** [comment]
FYI, a German proofreader told me he'd take a look in the evening. 

Update:

Feedback from a Danish translator on Crowdin:

"In danish we have æ, ø and å - and it differs how they are converted to slug. WP Permalink Manager for examples seems to convert to e,o,a: https://permalinkmanager.pro/docs/tutorials/special-characters-permalinks/ But on some websites i have also seen ae, oe, aa as a conversion."

Feedback from a German proofreader:

"ä -> ae
Ä -> Ae
ö -> oe
Ö -> Oe
ü -> ue
Ü -> Ue
ß -> ss

In case a word with such replaced character is shown in uppercase, it is completely fine to write both characters in capital letters (like UEBUNG instead of ÜBUNG, or STRASSE instead of STRAẞE).

In conclusion: **Native German speakers are absolutely used to replacing ä with ae etc.** in case we are not allowed or should not use umlauts (like for URLs). I'm pretty sure there are more characters like this in other languages, which have some special substitution like this."

**Comment by davidglezz** [comment]
In Spanish `ü -> ue` would be a incorrect transformation.

de_DE locale `ü -> ue`: `ÜBUNG → UEBUNG`
es_ES locale `ü -> u` : `pingüino → pinguino`  (not `pingueino`)

**Comment by wwatu1337** [comment]
In estonian we have õ,ä,ö,ü and currently the ae,oe,ue make no sense, maybe it should be a separate option?

Correct for estonian:
õ -> o
ä -> a
ö -> o
ü -> u
Õ -> O
Ä -> A
Ö -> O

**Comment by boo-code** [comment]
Measured this on 9.2.x. The situation has changed since the thread stalled, and @matks's question from 2021 now has a concrete answer.

**The PHP side is no longer a table.** #27176 was merged in April 2022 and replaced `replaceAccentedChars()` with ICU:

```php
$this->transliterator = Transliterator::create('Any-Latin; Latin-ASCII');
```

So `ö → o` is not an oversight in a list any more, it is what `Latin-ASCII` does: strip combining diacritics, but expand true ligatures and letters. That is why `ß → ss` and `æ → ae` already work while `ö` does not.

**The disagreement is exactly three characters.** Running the real `str2url()` from `js/admin.js` in node against `Tools::str2url()` over 41 accented characters:

| char | PHP (ICU) | JS (`js/admin.js`) |
|---|---|---|
| ä | a | ae |
| ö | o | oe |
| ü | u | ue |

Everything else agrees, including `ß → ss`, `æ → ae`, `œ → oe`, `ø → o`, `å → a`. So this is not a broad divergence, it is the three German umlauts, where `js/admin.js` still carries the pre-2022 hand-written expansion.

Two smaller observations from the same read. The JS lowercases before applying its tables (`toUpperCase()` then `toLowerCase()`), so the whole uppercase half never runs; and that half is itself inconsistent, since `Ä` is consumed by the `'A'` rule before the `'AE'` rule can see it, while `Ö` and `Ü` do reach `'OE'`/`'UE'`.

**On "what if some people expect 'o' and others 'oe'"** — that is exactly right, and ICU already encodes it per language:

```
de-ASCII                 Muesli Oel Kaese Strasse
Any-Latin; Latin-ASCII   Musli Ol Kase Strasse
```

`de-ASCII` reproduces the German proofreader's list in @Julievrz's comment exactly, including `ß → ss`. So the answer is not one global convention: German wants `oe`, Swedish and Finnish want `o` for the same characters, and the transliterator already knows which is which.

That leaves two directions, and both are a product call rather than mine:

1. **Align `js/admin.js` down to ICU** (`ä → a`, `ö → o`, `ü → u`). Three characters, one file, makes the two paths agree immediately. It follows the merged 2022 decision and the fact that the server is authoritative, since imports and the API never run the JS at all. The cost is that German shops lose `oe` from the back office button, which is the opposite of what this issue asked for.
2. **Pick the transliterator from the target language** — `de-ASCII` for German, `Any-Latin; Latin-ASCII` otherwise. This is the correct answer for every locale rather than a compromise, but `Tools::str2url()` takes no language today, so it needs threading through the call sites, and it changes generated URLs for existing German shops.

I have not prepared either, because the first contradicts the reporter and the second changes URL generation semantics shop-wide. Happy to build whichever is preferred — option 2 is the one I would argue for.
